sdk/trace/batch_span_processor_test.go

```diff
@@ -256,4 +255,9 @@ func TestBatchSpanProcessorShutdown(t *testing.T) {
 		t.Error("Error shutting the BatchSpanProcessor down\n")
 	}
 	assert.Equal(t, 1, bp.shutdownCount)
+
+	bspCopy, ok := bsp.(*batchSpanProcessor)
```

Comment:
This tests the internal state representation of being shutdown, but the PR is hoping to solve the problem where OnEnd and ForceFlush are called after shutdown and need to be noops. Can we add tests to check for that?

Reply:
My understanding was that the main work inside`ForceFlush` is `bsp.e.exportSpans` and it wouldn't do much after bsp gets shut down because it checks that `len(bsp.batch)>0` and at that point the queue doesn't get processed anymore so the batch would already be empty. I was hoping that I can use the `bsp.e == nil` to short circuit even earlier.

1) Do you see value in doing this check for early circuiting?

2) If so, I'm not sure how to test since the behavior would essentially be the same. The behavior might change if I move the timer reset to after the `bsp.e == nil` check and potentially do testing for something related to that (though I don't know what the best way would be)

Apologies if my understanding is off.

Reply:
Testing the overall behavior we want is going to provide more value. Testing that a span that is ended sent to the processor and then when the processor's `ForceFlush` is called never is delivered is the behavior we want this package to provide. Testing the internal implementation of how we do that will only be a good test as long as that implementation remains unchanged. As the package evolves and we updated that implementation we will still want the describe behavior, but we may do it in a different way. Writing tests that check this external behavior will therefore provide the project more value.